### ISSUE TITLE
libflux: fix module organization

### ISSUE THREAD

**Issue by affo** [influxdata/flux#2042]
In our Rust codebase, we create many files to conceptually separate pieces of code, but Rust automatically creates a module for a new file.

We should make multiple files be in the same module.

DOD: files in subfolders of libflux are in the same Rust module.

**Comment by nathanielc** [comment]
@affo Can you maybe be a bit more concrete with this issue? Maybe layout what the final directory structure should look like? And then also define what the module structure will look l like with examples of how to import and use the code from an external crate?

**Comment by affo** [comment]
@nathanielc I have updated the issue now.

**Comment by nathanielc** [comment]
@affo Let me make sure I follow:

Any new Rust source file is a new module. For that module to be visible to consumers of the crate we must use one of `pub mod` or `pub use <module path>`. The first exposes the module's public types as part of the _normal_ path (i.e. the directory and file names). The `pub use` expose the module's public types as if they were part of the module that has the `pub use` statement. In other words `pub use ` flattens the used module into the current module.

If I have `pub use <module>` do I also need the `mod <module>` statement somewhere to declare that the module exists?

**Comment by affo** [comment]
@nathanielc correct for the first part.

> If I have pub use <module> do I also need the mod <module> statement somewhere to declare that the module exists?

Yep, you need that for `cargo` to  understand your project structure, but that's not a problem, you can `mod <module` and `pub use <module>::...` to expose what you need to expose.

**Comment by nathanielc** [comment]
@affo thanks.

So what convention do we ant to use as a team?

 Should we treat directories as the _real_ module structure and treat each file as a private mod and then expose the public bits via the `pub use` in mod.rs for each directory?

Alternatively should we use less files to reduce the number of modules and keep the API  more closely mapped to the directory/file structure?

In general I dislike the indirection caused by creating lots of `pub use` statement because it becomes difficult to find the real source implementation of the API. 

Thoughts?

**Comment by affo** [comment]
I see you point on `pub use`.

Anyways, I really like clearly separating our libraries in multiple files, because I think it makes code easier to understand, review, and causes less conflicts.

So I would lean to modularizing more and introduce `pub use` in our `mod.rs`.

**Comment by jsternberg** [comment]
I think we should follow similar principles to how we would in Go. We should have ast, parser, and scanner as separate modules. From there, we would use our discretion for how we want to call the method. For example, I think `walk` should be exposed as a function in the ast module. But walk itself is pretty substantial so we should probably split it up into its own module. In that circumstance, we would have the implementation of walk in a hidden module, but we would potentially have `pub fn walk` inside of the AST module.

I don't think there are going to be hard guidelines on this, unfortunately. And some of it is going to involve some breaking changes while we figure out what organization structure is best. I do think that `pub use` is something that we should use. I don't know if we should create a team standard until we generally have more experience with Rust and I also don't think we should try to replicate the Go module structure we're used to.

**Comment by wolffcm** [comment]
I agree @affo's point here.  I think we should structure our module tree in a natural way that makes it easy for us to work on the code.  Then we should make use of the `pub` keyword, including `pub use` to present a coherent API to the user.  

In this case I really like the example of having `pub use ast::walk::walk`.  Consumers of the API would probably expect them to be in the same package.

The only issue I see is that within the package, code might refer to the same function via different paths in the code.  It might be good to come up with a standard way to handle that case (like always using the `pub` path if it exists).

**Comment by jlapacik** [comment]
+1 for `pub use` in mod files.

**Comment by nathanielc** [comment]
ok it seems we have a general consensus.

Give a module directory:

1. Use modules (aka files) when it logically makes sense to. Similar to how we do that with Go.
2. Import submodules privately into the parent module (i.e. using `mod <name>;`)
3. Export the public API from the parent modules using `pub use ` in `mod.rs` files.

**Comment by affo** [comment]
@nathanielc agreed 👍 

@wolffcm

> The only issue I see is that within the package, code might refer to the same function via different paths in the code.

I see your point, I created an example for that in the example repo: https://github.com/affo/learning-rust-modules/commit/708f5baabfb5e01e2c705ecaa218ec64cf156715.

Yeah, we should definitely pay attention to that. The compiler doesn't warn you about a double exported path.

**Comment by nathanielc** [comment]
I started playing with this to see if it makes sense. I found it didn't really make sense to re-export most things with `pub use` and rather found that `pub mod` worked better.

## AST Example

Take the `ast::walk` module for example. The public name exported from that module are: 

* walk - the walk function
* Node - the node enum to facilitate walking
* Visitor - the trait that defines how visitors should work
* create_visitor - a helper function to create a visitor from a closure

Using `pub mod` the API looks like:

```
use flux::ast;

ast::walk::walk
ast::walk::Node
ast::walk::Visitor
ast::walk::create_visitor
```

Consumers can choose to use `ast::walk` in order to make that shorter if desired:

```
use flux::ast::walk;

walk::walk
walk::Node
walk::Visitor
walk::create_visitor
```

Alternatively if we used `pub use crate::ast::walk::{walk,Node, Visitor,create_visitor};`

The API would look like:

```
use flux::ast;

ast::walk
ast::Node
ast::Visitor
ast::create_visitor
```

We loose the context of walking altogether. It also didn't make sense to only `pub use` part of the public API because all of the `ast::walk` names are important.


## Semantic Example

Take the `semantic` module as another example:

The current semantic module doesn't export anything public and all sub modules are private.

The current semantic/mod.rs:
```
mod analyze;
mod infer;
mod nodes;
mod sub;
mod types;
```

We probably want to keep the `analyze`, `infer`, and `sub` modules private and create a helper function `mod.rs` that does type inference from beginning to end without exposing the internal steps. This will naturally be public and exported as part of the semantic module API.

That leaves the `nodes` and `types` modules. They should likely be public.

Again there are two ways to make them public, `pub mod` or `pub use`.

In the `pub mod` case we would get:

```
// semantic/mod.rs
mod analyze;
mod infer;
pub mod nodes;
mod sub;
pub mod types;
```

```
use flux::semantic::nodes;
use flux::semantic::types;

nodes::Statement
nodes::Block
// etc

types::PolyType;
types::MonoType;
// etc
```


In the `pub use` case we would get:

```
// semantic/mod.rs
mod analyze;
mod infer;
mod nodes;
mod sub;
mod types;

pub use nodes::*;
pub use types::*;
```

```
use flux::semantic;
use flux::semantic;

semantic::Statement
semantic::Block
// etc

semantic::PolyType;
semantic::MonoType;
// etc
```

The `nodes` `types` organization felt helpful to me and it didn't get in the way. 

## Partial Pub use

In no case did I find an example where doing a partial `pub use` was a clear win. By _partial pub use_ I mean using only a subset of the module. For example doing `pub use ast::walk::walk;`, Importing `walk` without its `Visitor` and `Node` types seemed wrong. So it was an all or nothing approach.


## Conclusion

Since it became an all or nothing approach there isn't much difference between `pub mod` and `pub use` other than a single extra name and in most cases I found that name helpful. And I found that `pub mod` meant that there was less indirection which makes it easier to find the code.

Using `pub mod` also eliminates the confusion over there being multiple paths for importing the same type.


My recommendation therefore is we can have a mix of public and private sub-modules but we should use `pub mod` to expose sub-modules as appropriate. Leveraging `pub use` should be the exception.

**Comment by affo** [comment]
@nathanielc Thanks for this.

Clear and detailed explanation. I think that matches what developers of Rust do. For example `std::result::Result` or `std::cell::Cell`.

As long as we can keep `pub mod` let's go for that. If (and only if) or internal structure could be crazy un unusable from the outside, then we should expose the bits with `pub use`. As you highlighted above, our current organization of modules lets us expose functions via `pub mod` making sense to the final user. So, let's go for it.

I guess this main purpose, at this point, is to raise awareness around `pub mod` VS `pub use` and that we can use them.

**Comment by nathanielc** [comment]
Here is a PR https://github.com/influxdata/flux/pull/2083 that makes the above changes if it helps understand the structure I am proposing. (its a small change from what we have now).

**Comment by timhallinflux** [comment]
Seems like this should be reviewed and closed?
